mixnet/client/src/sender.rs

```diff
@@ -0,0 +1,106 @@
+use std::{error::Error, net::SocketAddr};
+
+use mixnet_topology::Topology;
+use nym_sphinx::{
+    addressing::nodes::NymNodeRoutingAddress, chunking::fragment::Fragment, message::NymMessage,
+    params::PacketSize, Delay, Destination, DestinationAddressBytes, NodeAddressBytes,
+    IDENTIFIER_LENGTH, PAYLOAD_OVERHEAD_SIZE,
+};
+use rand::rngs::OsRng;
+use sphinx_packet::{route, SphinxPacket, SphinxPacketBuilder};
+use tokio::{io::AsyncWriteExt, net::TcpStream};
+
+// Sender splits messages into Sphinx packets and sends them to the Mixnet.
+pub struct Sender {
+    //TODO: handle topology update
+    topology: Topology,
+}
+
+impl Sender {
+    pub fn new(topology: Topology) -> Self {
+        Self { topology }
+    }
+
+    pub fn send(
+        &self,
+        msg: Vec<u8>,
+        destination: SocketAddr,
+        num_hops: usize,
+    ) -> Result<(), Box<dyn Error>> {
+        let dest_addr: NodeAddressBytes =
+            NymNodeRoutingAddress::from(destination).try_into().unwrap();
+        let destination = Destination::new(
+            DestinationAddressBytes::from_bytes(dest_addr.as_bytes()),
+            [0; IDENTIFIER_LENGTH], // TODO: use a proper SURBIdentifier if we need SURB
+        );
+
+        let mut packets: Vec<(sphinx_packet::SphinxPacket, route::Node)> = Vec::new();
+
+        for fragment in Sender::pad_and_split_message(msg) {
+            let packet = self.build_sphinx_packet(fragment, &destination, num_hops)?;
+            packets.push(packet);
+        }
+
+        for (packet, first_node) in packets {
+            tokio::spawn(async move {
+                if let Err(e) = Sender::send_packet(Box::new(packet), first_node.address).await {
+                    tracing::error!("failed to send packet to the first node: {e}");
+                }
+            });
```

Comment:
nit: `Sender::pad_and_split_message(msg).map(|..| build_sphinx_packet(..)).collect::<Result<_>>()?` (yes you can go from an iterator of results into a result of a collection) and maybe even `.for_each(tokio::spawn(..))`

Reply:
thank you: 20be7cf